src/libcore/iter.rs

```diff
@@ -1022,6 +1021,47 @@ pub trait Iterator {
             }
         }
     }
+
+    /// Iterates over the entire iterator, summing up all the elements
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// # #![feature(core)]
+    ///
+    /// let a = [1, 2, 3, 4, 5];
+    /// let mut it = a.iter().cloned();
+    /// assert!(it.sum() == 15);
+    /// ```
+    #[unstable(feature="core")]
+    fn sum<T, S=T>(self) -> S where
```

Comment:
Could this also omit the `T` type parameter? I think something like this in theory shouldn't ICE:

```rust
fn sum<S=Self::Item>(self) -> S
    where S: Add<Self::Item,Output=S> + Zero, Self: Sized
{
     ...
}
```